app/src/main/kotlin/com/github/gotify/messages/MessagesActivity.kt

```diff
@@ -0,0 +1,638 @@
+        }
+    }
+
+    private inner class MessageListOnScrollListener : RecyclerView.OnScrollListener() {
+        override fun onScrollStateChanged(view: RecyclerView, scrollState: Int) {}
+
+        override fun onScrolled(view: RecyclerView, dx: Int, dy: Int) {
+            val linearLayoutManager = view.layoutManager as LinearLayoutManager?
+            if (linearLayoutManager != null) {
+                val lastVisibleItem = linearLayoutManager.findLastVisibleItemPosition()
+                val totalItemCount = view.adapter!!.itemCount
+                if (lastVisibleItem > totalItemCount - 15 &&
+                    totalItemCount != 0 &&
+                    viewModel.messages.canLoadMore(viewModel.appId)
+                ) {
+                    if (!isLoadMore) {
+                        isLoadMore = true
+                        launchCoroutine {
+                            loadMore(viewModel.appId)
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private suspend fun updateMissedMessages(id: Long) {
+        if (id == -1L) return
+
+        val newMessages = MissedMessageUtil(viewModel.client.createService(MessageApi::class.java))
+            .missingMessages(id).filterNotNull()
+        viewModel.messages.addMessages(newMessages)
+
+        if (newMessages.isNotEmpty()) {
+            updateMessagesForApplication(true, viewModel.appId)
+        }
+    }
+
+    override fun onCreateOptionsMenu(menu: Menu): Boolean {
+        menuInflater.inflate(R.menu.messages_action, menu)
+        menu.findItem(R.id.action_delete_app).isVisible =
+            viewModel.appId != MessageState.ALL_MESSAGES
+        return super.onCreateOptionsMenu(menu)
+    }
+
+    override fun onOptionsItemSelected(item: MenuItem): Boolean {
+        if (item.itemId == R.id.action_delete_all) {
+            launchCoroutine {
+                deleteMessages(viewModel.appId)
+            }
+        }
+        if (item.itemId == R.id.action_delete_app) {
+            val alert = android.app.AlertDialog.Builder(this)
+            alert.setTitle(R.string.delete_app)
+            alert.setMessage(R.string.ack)
+            alert.setPositiveButton(R.string.yes) { _, _ -> deleteApp(viewModel.appId) }
+            alert.setNegativeButton(R.string.no, null)
+            alert.show()
+        }
+        return super.onContextItemSelected(item)
+    }
+
+    private fun deleteApp(appId: Long) {
+        val settings = viewModel.settings
+        val client = ClientFactory.clientToken(settings.url, settings.sslSettings(), settings.token)
+        client.createService(ApplicationApi::class.java)
+            .deleteApp(appId)
+            .enqueue(
+                Callback.callInUI(
+                    this,
+                    onSuccess = { refreshAll() },
+                    onError = { Utils.showSnackBar(this, getString(R.string.error_delete_app)) }
+                )
```

Comment:
Yeah, that's because the Callback relies on non-null responses.
Maybe I misunderstood as you said in one of the comments above.

> The gotify server should never return a null response, so its fine to throw here.
(https://github.com/gotify/android/pull/258#discussion_r1061786262)

But it seems the delete call does not return anything, so we have to change that again.

Reply:
True, I meant my comment a little different: Gotify never returns null, when there should be a response. We can add extra handling when we expect a Void response:
To `Api.kt` add:
```
    @Throws(ApiException::class)
    fun execute(call: Call<Void>) {
        try {
            val response = call.execute()
            if (!response.isSuccessful) {
                throw ApiException(response)
            }
        } catch (e: IOException) {
            throw ApiException(e)
        }
    }
```
this will fix most of the delete calls, then there is one delete call which uses Callback.call(), we can add the following methods, to callback
```kotlin
        fun <T> call(onSuccess: ResultSuccessCallback<T> = ResultSuccessCallback {}, onError: ErrorCallback = ErrorCallback {}): retrofit2.Callback<T> {
            return RetrofitCallback(merge(of(onSuccess, onError), errorCallback()))
        }
        
        fun <T> callVoid(onSuccess: VoidSuccessCallback<T> = VoidSuccessCallback {}, onError: ErrorCallback = ErrorCallback {}): retrofit2.Callback<T> {
            return RetrofitCallback(merge(of(onSuccess, onError), errorCallback()))
        }

        private fun <T> of(onSuccess: SuccessCallback<T>, onError: ErrorCallback): Callback<T> {
            return Callback(onSuccess, onError)
        }
```
and change the interface definitions to this:
```kotlin
    fun interface SuccessCallback<T> {
        fun onSuccess(body: T?, response: Response<T>)
    }

    fun interface ResultSuccessCallback<T> : SuccessCallback<T> {
        override fun onSuccess(body: T?, response: Response<T>) {
            onResultSuccess( body ?: throw ApiException("null response", response))
        }

        fun onResultSuccess(data: T)
    }

    fun interface VoidSuccessCallback<T> : SuccessCallback<T> {
        override fun onSuccess(body: T?, response: Response<T>) {
            onVoidSuccess()
        }

        fun onVoidSuccess()
    }

    fun interface ErrorCallback {
        fun onError(t: ApiException)
    }

    private class RetrofitCallback<T>(private val callback: Callback<T>) : retrofit2.Callback<T> {
        override fun onResponse(call: Call<T>, response: Response<T>) {
            if (response.isSuccessful) {
                callback.onSuccess.onSuccess(response.body(), response)
            } else {
                callback.onError.onError(ApiException(response))
            }
        }

        override fun onFailure(call: Call<T>, t: Throwable) {
            callback.onError.onError(ApiException(t))
        }
    }
```
what do you think?

Reply:
I've copy-pasted your code and found the implementation has some problems. I would...

... agree with the first part except for the type 'Void', I would use the Kotlin equivalent 'Unit' (it should work the same way):
``` kotlin
@Throws(ApiException::class)
fun execute(call: Call<Unit>)
```

... also agree with the second part, if there wouldn't be e.g. the `callInUI` function (which I think should accept the ResultSuccessCallback as well).
Because it runs the code `context.runOnUiThread { onSuccess.onSuccess(data) }` it calls the onSuccess function of the callback. Unfortunately you've extended the function signature to `fun onSuccess(body: T?, response: Response<T>)`.

So the `callInUI` function throws an IDE error telling us to also hand over the `reponse` parameter, which we don't have access to here, because we only hand over the body to the lambda. Same problem in the `merge` function.

So I would simply do this:

``` kotlin
fun interface ResultSuccessCallback<T> : SuccessCallback<T> {
    override fun onSuccess(body: T?, response: Response<T>) {
        onResultSuccess(body ?: throw ApiException("null response", response), response)
    }

    fun onResultSuccess(data: T, response: Response<T>)
}
```

Now we can call 

``` kotlin
onSuccess = { data, response ->
    context.runOnUiThread { onSuccess.onSuccess(data, response) }
}
```

 in the `callInUI` function.

And when implementing we would of course change the occurences of `SuccessCallback` to `ResultSuccessCallback`, except for the delete functions.

What do you think?

Reply:
I don't really like having the response in the ResultSuccessCallback, it shouldn't be needed, no consumer of the not null onResultSuccess method needs this parameter.

We could accept the `SuccessCallback<T>` interface on callInUI and then specify which callback is required on the caller side, e.g.
```
Callback.callInUI(
    this,
    onSuccess = Callback.ResultSuccessCallback { client -> onCreatedClient(client) },
    onError = { onFailedToCreateClient() }
)
```

Reply:
That wouldn't solve the situation I think. Here's a quick explanation of what I mean:

- the interface SuccessCallback provides the function `onSuccess` with body and response as parameters
- the callInUI function has a ResultSuccessCallback
- we call the function `onSuccess` on the given callback, but we don't have a value for `response` (see response highlighted red)
- that's because we only provide data to the onVoidSuccess of the callback (see blue arrow)

<img width="547" alt="image" src="https://user-images.githubusercontent.com/46033730/212563977-ef8ad472-38a0-4a06-9b07-9d3b2c65febc.png">

---

With your suggestion we would still hand over a SuccessCallback, which again wouldn't have a reponse included to call `onSuccess` internally.

Or did I got you wrong?

Reply:
If Callback.call takes a SuccessCallback then you can write it like this:
```
        fun <T> callInUI(
            context: Activity,
            onSuccess: SuccessCallback<T>,
            onError: ErrorCallback
        ): retrofit2.Callback<T> {
            return call(
                onSuccess = { data, response -> context.runOnUiThread { onSuccess.onSuccess(data, response) } },
                onError = { exception -> context.runOnUiThread { onError.onError(exception) } }
            )
        }


        fun <T> call(onSuccess: SuccessCallback<T>, onError: ErrorCallback): retrofit2.Callback<T> {
            return RetrofitCallback(merge(of(onSuccess, onError), errorCallback()))
        }
```
because the SuccessCallback will be implemented which has the onSuccess(data: T?, response) method.

Reply:
This forces you to always specify which concrete implementation of the success callback to use, like described in my last comment.

Reply:
Ah, I see.
Let's see if I can get it working.

Reply:
So I've decided to leave these last bits of implementation to you.

1. More functions showed up that need extra handling like `lastReceivedMessage` of the MissedMessageUtil and I don't exactly know how you want this to be implemented, so I give that choice to you as the project owner.
2. I don't want to "harm" your code in case of writing my own style of Callbacks you probably don't want here.

PS: Here are the functions I've already adapted: https://gist.github.com/cyb3rko/85305469a1d2229da356bef6f243096b
PPS: Went back to using Void instead of Unit, I realized it all depends on the Java API client which uses Void.
PPPS: I will try to enjoy the last hours of my brithday xD, I'm out

Reply:
Alright, I'm busy until next week, so it'll take some time.

Reply:
I've fixed this problem, could you have a look over the new code? I'd be okay with merging it, but there are currently merge conflicts with master.